### PR TITLE
Add codebase init command, no longer create codebases implicitly

### PR DESCRIPTION
**Pull request by sullyj3**
~Questions:
Can I use lambdacase in main? (Main.hs:118)
How can I get rid of the parens around the do block in FileCodebase? (FileCodebase.hs:125)
Is it OK to pull out formatAnn to the top level? I'm not sure what it's for. (FileCodebase.hs:155)~

### CODE REVIEW

**Comment by pchiusano** [comment]
Hi @sullyj3 we ended up needing to shuffle a bunch of things around in `Main`, sorry about that! Would you mind updating this PR to a) add yourself to CONTRIBUTORS file and b) work against master. Here is an updated spec for how it should work:

* `ucm init` creates a new codebase in the user's home directory, (so `~/.unison`), then exits with a nice message "I successfully created a new codebase at <path>" or something. If there's already a codebase there, it prints a friendly message: "I didn't initialize a new codebase at <path> because there is an existing Unison codebase at that location."
* `ucm -codebase /some/path init` creates a codebase in `/some/path/.unison`, then exits. Same behavior if there's already a codebase there.
* Other subcommands should exit with a nice error message like "I didn't find a Unison codebase at the location <path>. Run `ucm -codebase <path> init` [or `ucm init` if user didn't supply `-codebase`] to initialize one, then try again." Notice how this pattern avoids a modal input, but the workflow is still quite efficient - user types `ucm`, then hits up arrow, then types `init<Enter>`. In general, we try to do everything without any modal inputs unless there's a compelling reason not to. Anytime you'd think of doing something with a modal input, just exit with a message about how to complete the command, rather than modally forcing the user to complete the command (and then you need a way to cancel, etc).
* `help` and `version` subcommands don't need a codebase.

You can see how the existing `Main` deals with the `-codebase` flag. Basically, once you have a location for where the codebase should be, you can do a check if one exists there (might need a helper function `FileCodebase.existsAt : FilePath -> IO Bool`).

Sorry for the churn! If you are out of steam on this let us know and someone else could take this over. Cheers!

**Comment by pchiusano** [comment]
One other thing - we also discussed using opt-parse-applicative. I'm not super happy with the help it autogenerates or the error messages it produces, though maybe that can be customized. We could also just use Megaparsec if the ucm argumets / subcommands get more complicated (I really think we should try to keep them simple though). In any case, I'd leave that as a possible refactoring for later and stick with the simple arg parsing via pattern matching that's there right now.

**Comment by sullyj3** [comment]
Cool cool, so `-codebase` must be supplied (if it is supplied at all) before any subcommands?

**Comment by pchiusano** [comment]
@sullyj3 yes that’s right. That -codebase flag already exists and you can see how it’s done if you check current Main in master.

**Comment by sullyj3** [comment]
Ok, seems to mostly be working. Couple of questions:

Firstly, I can't seem to get the Pretty printing to separate lines properly, I'm getting messages that look like:
```No codebase exists at /home/james Run `ucm init` to create one, then try again!```
when I expect `Run` onwards to be on a new line.

 Any ideas?

Secondly, I tried `unison transcript.fork hello.md`. It seemed to terminate successfully:
```
  I've finished running the transcript(s) in this codebase:

    /home/james/Code/sullyj3_unison/unison-src/transcript-b4079f501b0b2d89

  You can run
 `ucm -codebase /home/james/Code/sullyj3_unison/unison-src/transcript-b4079f501b0b2d89` to do more work with it.
```
I followed the suggestion
`unison -codebase transcript-b4079f501b0b2d89`
And it seemed to load successfully. However...
```
.> view x

  ⚠️

  The following names were not found in the codebase. Check your spelling.
    x
```

Is this expected?

**Comment by sullyj3** [comment]
Oops, will get those tests passing

**Comment by sullyj3** [comment]
Fixed it. In the process I altered the transcript runner to use `.unison/v1` codebases rather than just `.unison`, let me know if that's not right.

File: parser-typechecker/unison/Main.hs [hunk 1]
```diff
@@ -63,60 +63,71 @@ usage = P.callout "🌻" $ P.lines [
...
+  args <- getArgs
   -- hSetBuffering stdout NoBuffering -- cool
-  (codepath, args) <- case args of
-    "-codebase" : codepath : args -> pure (codepath, args)
-    _ -> (,) <$> getHomeDirectory <*> pure args
+
+  -- We need to know whether the program was invoked with -codebase for
+  -- certain messages. Therefore we keep a Maybe FilePath - mcodepath
+  -- rather than just deciding on whether to use the supplied path or
+  -- the home directory here and throwing away that bit of information
```
Comment (aryairani) [inline parser-typechecker/unison/Main.hs]: Thanks for including this comment.  I was about to ask, and then I thought "what if I try reading..."

**Comment by aryairani** [comment]
Oops, let me take a look at your question I just now spotted.

**Comment by aryairani** [comment]
Re. the pretty-printing: that's odd and I'm not sure.  You've got the `P.lines` call in there, which I believe should be enough.  I might be indicative of a bug related to #820?

And re. https://github.com/unisonweb/unison/pull/826#issuecomment-537510938 did I understand correctly that 1a6ab46 fixed the issue you were seeing?

**Comment by aryairani** [comment]
And yes, feel free to use LambdaCase wherever you like.

**Comment by sullyj3** [comment]
> And re. [#826 (comment)](https://github.com/unisonweb/unison/pull/826#issuecomment-537510938) did I understand correctly that [1a6ab46](https://github.com/unisonweb/unison/commit/1a6ab46a159f30cf402d76edcd86387a3a59afb6) fixed the issue you were seeing?

No, sorry, that's a separate thing I'm still confused about. Is it something to do with needing to qualify the name, maybe?

**Comment by sullyj3** [comment]
> Re. the pretty-printing: that's odd and I'm not sure. You've got the `P.lines` call in there, which I believe should be enough. It might be indicative of a bug related to #820?

Seems likely, but also, the version I have slightly above using `<>` and `P.newline`, rather than P.lines has the same issue.

**Comment by pchiusano** [comment]
So aside from the formatting thing, this is good to go? If so I’m inclined to merge and then investigate the formatting later.

**Comment by sullyj3** [comment]
I think so. Sounds good.

**Comment by pchiusano** [comment]
Yay! Thanks @sullyj3. I’ll work on tracking down the weird formatting issue.

**Pull request merged**